### DISCUSSION TITLE
Low hardware tips and fine tunnings

### DISCUSSION THREAD

**Discussion by sorinpenteleiciuc** [RomanIakovlev/timeshape#102] [Q&A]
I am experiencing performance problems when calling 
` TimeZoneEngine.initialize();`
Because my machine is very low hardware ( like an old Raspberry Pi )

Im also saw this question: https://github.com/RomanIakovlev/timeshape/discussions/73 so I would like to maybe do some fine tunnings like setting the **accelerateGeometry** to true. Maybe I could sacrifice some precision.

**Comment by RomanIakovlev** [comment]
I think it's awesome that you try to run Timeshape on such a miniature device! I'd be really interested to get to the bottom of the problem.

I'd say you should first try initializing it with a smaller bounding box plus *disable* `accelerateGeometry`. You should use this method: https://github.com/RomanIakovlev/timeshape/blob/master/core/src/main/java/net/iakovlev/timeshape/TimeZoneEngine.java#L233. 

Keep in mind the time zone should be completely covered with the bounding box you provide, that is, it should entirely inside the bounding box, not just overlap with it. Otherwise that zone will not be included in the index, potentially leaving it empty so that no queries would work at all.

Regarding `accelerateGeometry`, it makes the queries faster, but it takes longer to build the index with it enabled. Precision is not affected. If you have issues with initializing, you should disable it at first.

**Reply by sorinpenteleiciuc** [reply]
Could you please give me a hand how I could limit the perimeter to Europe ? could I try Europe and if the query is negative then try South America ? I only need actually to do only 1 single query and not multiple queries.

**Reply by RomanIakovlev** [reply]
For the whole mainland Europe (without England and Scandinavia) you could try these coordinates: minLat = 34.651, minLon = -13.623, maxLat = 60.229, maxLon = 28.521. The South America would be roughly -56.99, -82.62, 14.86, -32.34.

By the way, what exactly is the problem you're facing with initializing? It just never finishes, or throws an error, or something else?

**Reply by sorinpenteleiciuc** [reply]
The error I receive, is an out of memory.

Nov 15 08:28:44 kernel: Out of memory: Kill process 1579 (java) score 751 or sacrifice child

Do you have any tips ?

**Reply by RomanIakovlev** [reply]
How much total RAM do you have on that device? Maybe you could try to specify the Java heap and stack explicitly when you run your app, using the `java -Xmx` approach, so that it fits into the available memory?

**Reply by sorinpenteleiciuc** [reply]
I have around 512 MEGA for the whole device. I start with java -jar -Xms256m -Xmx512m. I wanted to ask also about calling the garbage collector after maybe doing that 1 query that I need.

**Reply by RomanIakovlev** [reply]
I think you could try to set both `-Xms` and `-Xmx` to the same, lower value, something like `150m`. Timeshape needs ~128 MB to initialize the whole world, see https://github.com/RomanIakovlev/timeshape/blob/master/doc/Architecture.md#memory-usage. 

So, call your program something like this:
```
java -jar -Xms150m -Xmx150m
```
If 150m is not sufficient, e.g. if your program, besides Timeshape data, needs more than 150-128=22m, you could try to gradually raise that 150m limit. But in general, you should not specify the value of max heap equal to your whole amount of RAM, that's surely triggering the `oom_killer`. 

And that 128 MB number is for the whole world, if you only load a subset that number will be lower.
